### PR TITLE
Support surrogate pairs

### PR DESCRIPTION
**Pull request by falsandtru**

### CODE REVIEW

File: implementation.js [hunk 1]
```diff
@@ -157,6 +157,8 @@ module.exports = function from(arrayLike) {
 	if (arrayFromIterable) {
 		items = arrayFromIterable;
 		len = arrayFromIterable.length;
+	} else if (typeof arrayLike === 'string') {
+		return items.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]?|[^\uD800-\uDFFF]|./g) || [];
```
Comment (ljharb) [inline implementation.js]: this should not happen unless it's in an environment without `Symbol.iterator` - how can we ensure that? If the environment has that, we must look up the `Symbol.iterator` function off of the string, and obey it.

**Comment by falsandtru** [comment]
This is created for older envs that have not `Symbol.iterator`. Tests that I added would be failed if I don't add the surrogate pairs support.

**Comment by ljharb** [comment]
In which browsers have you tested this? Can you also add a test (skipped when appropriate) that makes a String object, and sets `Symbol.iterator` on it to do something different, that ensures that it obeys that instead of iterating the code points?

File: implementation.js (hunk 1 above)
Comment (falsandtru) [inline implementation.js]: When strings are iterable, they must be parsed by `arrayFromIterable` function. So string splitting using the iterator must be ensured in that function. In this code, my added code will be skipped when strings are iterable.

**Comment by falsandtru** [comment]
Okay. Another, now I don't support string-like objects such as `Object('abc')` and `new String('abc')`. How should I do?

**Comment by ljharb** [comment]
We should use https://npmjs.com/is-string if we want to detect both primitive and object strings.

**Comment by falsandtru** [comment]
Thanks, I'll use it.

File: tests/tests.js [hunk 2]
```diff
@@ -75,6 +75,17 @@ var runTests = function runTests(arrayFrom) {
 	test('works with strings', function (t) {
 		t.deepEqual(arrayFrom(''), []);
 		t.deepEqual(arrayFrom('abc'), 'abc'.split(''));
+		t.deepEqual(arrayFrom('a\nb\nc\n\n'), 'a\nb\nc\n\n'.split(''));
+		t.deepEqual(arrayFrom('foo\uD834\uDF06bar'), ['f', 'o', 'o', '\uD834\uDF06', 'b', 'a', 'r']);
```
Comment (ljharb) [inline tests/tests.js]: Could we add tests for all these strings that use the object form as well?

Comment (falsandtru) [inline tests/tests.js]: Yes, I added.

File: tests/tests.js [hunk 3]
```diff
@@ -72,9 +72,30 @@ var runTests = function runTests(arrayFrom) {
...
+		t.deepEqual(arrayFrom(Object('abc')), 'abc'.split(''));
+		t.deepEqual(arrayFrom(Object('a\nb\nc\n\n')), 'a\nb\nc\n\n'.split(''));
+		t.deepEqual(arrayFrom(Object('foo\uD834\uDF06bar')), ['f', 'o', 'o', '\uD834\uDF06', 'b', 'a', 'r']);
+		t.deepEqual(arrayFrom(Object('foo\uD834bar')), ['f', 'o', 'o', '\uD834', 'b', 'a', 'r']);
+		t.deepEqual(arrayFrom(Object('foo\uDF06bar')), ['f', 'o', 'o', '\uDF06', 'b', 'a', 'r']);
+		t.end();
+	});
+
+	test('ensure to use iterators with strings', { 'skip': typeof Symbol !== 'function' }, function (t) {
+		t.deepEqual(arrayFrom('foo\uD834\uDF06bar'), Array.from('foo\uD834\uDF06bar'));
```
Comment (ljharb) [inline tests/tests.js]: we should also, in this test, have something like:
```js
var str = Object('foo\uD834\uDF06bar');
const arr = [1, 2, 3];
str[Symbol.iterator] = function () { return arr.values(); };
t.deepEqual(arrayFrom(str), arr);
```

Comment (falsandtru) [inline tests/tests.js]: Seems like `[].values()` is still not supported with Node 6.5.0. Do we really use that method?

Comment (falsandtru) [inline tests/tests.js]: How is this?

```js
	test('ensure to use iterators with strings', { 'skip': typeof Symbol !== 'function' }, function (t) {
		var a = Object('a');
		var b = Object('b');
		a[Symbol.iterator] = function () { return b[Symbol.iterator](); };
		t.deepEqual(arrayFrom(a), ['b']);
		t.end();
	});
```

Comment (falsandtru) [inline tests/tests.js]: > Keys or entries

ok.

Comment (falsandtru) [inline tests/tests.js]: Ah, unfortunately, those methods are not existed in strings. Can we use the default iterator and the test that I pasted?

Comment (ljharb) [inline tests/tests.js]: Right, that's why I suggested making the string iterate over `[1, 2, 3]` - we don't need to actually iterate the string, we just need to ensure that the iterator returned by the `Symbol.iterator` function is used.

Comment (ljharb) [inline tests/tests.js]: More specifically, we must *not* iterate over the string, in order to verify that.

Comment (falsandtru) [inline tests/tests.js]: Do we need any more changes?

File: implementation.js [hunk 4]
```diff
@@ -157,6 +158,8 @@ module.exports = function from(arrayLike) {
 	if (arrayFromIterable) {
 		items = arrayFromIterable;
 		len = arrayFromIterable.length;
+	} else if (isString(items)) {
+		return items.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]?|[^\uD800-\uDFFF]|./g) || [];
```
Comment (ljharb) [inline implementation.js]: To be extra robust, we should cache `String.prototype.match` at the module level, and `match.call(items, regex)` here.

Comment (falsandtru) [inline implementation.js]: I see, which do you like better?

```js
return String.prototype.match.call(items, /[\uD800-\uDBFF][\uDC00-\uDFFF]?|[^\uD800-\uDFFF]|./g) || [];
```

```js
return ''.match.call(items, /[\uD800-\uDBFF][\uDC00-\uDFFF]?|[^\uD800-\uDFFF]|./g) || [];
```

Comment (falsandtru) [inline implementation.js]: sorry, removed the previous comment.

Comment (falsandtru) [inline implementation.js]: I see, fixed.

**Comment by ljharb** [comment]
Thanks!

**Pull request merged**